### PR TITLE
feat: Add spanId attribute for span-log join query

### PR DESCRIPTION
**Pull request by rish691**
hypertrace/hypertrace#224

### CODE REVIEW

File: hypertrace-core-graphql-span-schema/src/main/java/org/hypertrace/core/graphql/span/dao/SpanLogEventRequestBuilder.java [hunk 1]
```diff
@@ -29,31 +33,67 @@
...
+  private Single<Collection<AttributeRequest>> getRequestAttributes(
+      GraphQlRequestContext requestContext, Collection<AttributeRequest> logEventAttributes) {
+    return this.attributeStore
+        .getForeignIdAttribute(
+            requestContext,
+            HypertraceCoreAttributeScopeString.LOG_EVENT,
+            HypertraceCoreAttributeScopeString.SPAN)
+        .flatMap(
+            spanId ->
+                logEventAttributes.stream()
```
Comment (aaron-steinfeld) [inline hypertrace-core-graphql-span-schema/src/main/java/org/hypertrace/core/graphql/span/dao/SpanLogEventRequestBuilder.java:84]: Since we're returning a collection anyway, can we just collect to a set rather than check if we already have it or not?

Comment (aaron-steinfeld) [reply hypertrace-core-graphql-span-schema/src/main/java/org/hypertrace/core/graphql/span/dao/SpanLogEventRequestBuilder.java:84]: We also could 1. probably simplify the nesting by just merging the two sources and 2. remove the nesting above by converting the attributes before returning. Something like:
```java
this.attributeStore
  .getForeignIdAttribute(
      requestContext,
      HypertraceCoreAttributeScopeString.LOG_EVENT,
      HypertraceCoreAttributeScopeString.SPAN)
  .map(attributeRequestBuilder::buildForAttribute) // can just inject this builder rather than create our own impl
  .toObservable()
  .mergeWith(Observable.fromIterable(logEventAttributes))
  .collect(Collectors.toUnmodifiableSet())
  .flatMap(this.attributeConverter::convert)
 ```

File: hypertrace-core-graphql-span-schema/src/main/java/org/hypertrace/core/graphql/span/dao/SpanLogEventResponseConverter.java [hunk 2]
```diff
@@ -54,18 +54,37 @@
       LogEventsResponse logEventsResponse) {
     return Observable.fromIterable(logEventsResponse.getLogEventsList())
         .concatMapSingle(
-            logEventsResponseVar -> this.convert(attributeRequests, logEventsResponseVar))
-        .collect(Collectors.groupingBy(logEvent -> (String) logEvent.attribute(foreignIdAttribute)))
+            logEventsResponseVar ->
```
Comment (aaron-steinfeld) [inline hypertrace-core-graphql-span-schema/src/main/java/org/hypertrace/core/graphql/span/dao/SpanLogEventResponseConverter.java:57]: What's the motivation for the change here? Since we were assuming the attribute existed in the response map and now we're guaranteeing it, the old logic would have also worked, right?

Comment (rish691) [reply hypertrace-core-graphql-span-schema/src/main/java/org/hypertrace/core/graphql/span/dao/SpanLogEventResponseConverter.java:57]: If the original for logs didn't have the spanId attribute we shouldn't return it back either. So, this part needed changes to filter out the spanId returned if it wasn't requested.

Comment (aaron-steinfeld) [reply hypertrace-core-graphql-span-schema/src/main/java/org/hypertrace/core/graphql/span/dao/SpanLogEventResponseConverter.java:57]: based on the way the response object is designed, it's safe to return extra data (because it's not a programmatic access but instead via the schema, we already know at request time what values of the map will be read by the serializer - that's how we generated the requested attributes)

File: hypertrace-core-graphql-span-schema/src/main/java/org/hypertrace/core/graphql/span/dao/SpanLogEventRequestBuilder.java [hunk 3]
```diff
@@ -29,31 +33,67 @@
...
-                    gqlRequest.spanEventsRequest().timeRange().startTime().toEpochMilli())
-                .setEndTimeMillis(
-                    gqlRequest.spanEventsRequest().timeRange().endTime().toEpochMilli())
-                .addAllSelection(selections)
-                .setFilter(filter)
-                .build());
+    return getRequestAttributes(
+            gqlRequest.spanEventsRequest().context(), gqlRequest.logEventAttributes())
+        .flatMap(
+            v ->
```
Comment (skjindal93) [inline hypertrace-core-graphql-span-schema/src/main/java/org/hypertrace/core/graphql/span/dao/SpanLogEventRequestBuilder.java:55]: Can we please have a more descriptive name for `v`? It could even say `attribute` or `requestAttribute` or whatever best describes it

File: hypertrace-core-graphql-span-schema/src/main/java/org/hypertrace/core/graphql/span/dao/SpanLogEventRequestBuilder.java [hunk 4]
```diff
@@ -29,31 +33,67 @@
...
+                                logEventAttribute.attribute().key().equals(spanId.key()))
+                    ? Single.just(logEventAttributes)
+                    : updateRequestAttribute(spanId, logEventAttributes));
+  }
+
+  private Single<Collection<AttributeRequest>> updateRequestAttribute(
+      AttributeModel attributeModel, Collection<AttributeRequest> requests) {
+    List<AttributeRequest> list = new ArrayList<>(requests);
+    list.add(new DefaultAttributeRequest(attributeModel, ""));
+    return Single.just(list);
```
Comment (skjindal93) [inline hypertrace-core-graphql-span-schema/src/main/java/org/hypertrace/core/graphql/span/dao/SpanLogEventRequestBuilder.java:96]: I am guessing this is a mutable list. Can we convert it into an unmodifiable list? Basically, you can use `Collections.unmodifiableList(list)`

File: hypertrace-core-graphql-span-schema/src/main/java/org/hypertrace/core/graphql/span/dao/SpanLogEventRequestBuilder.java [hunk 5]
```diff
@@ -29,31 +33,67 @@
...
+                            logEventAttribute ->
+                                logEventAttribute.attribute().key().equals(spanId.key()))
+                    ? Single.just(logEventAttributes)
+                    : updateRequestAttribute(spanId, logEventAttributes));
+  }
+
+  private Single<Collection<AttributeRequest>> updateRequestAttribute(
+      AttributeModel attributeModel, Collection<AttributeRequest> requests) {
+    List<AttributeRequest> list = new ArrayList<>(requests);
+    list.add(new DefaultAttributeRequest(attributeModel, ""));
```
Comment (skjindal93) [inline hypertrace-core-graphql-span-schema/src/main/java/org/hypertrace/core/graphql/span/dao/SpanLogEventRequestBuilder.java:95]: An empty alias string 🤔. When is that expected?

File: hypertrace-core-graphql-span-schema/src/main/java/org/hypertrace/core/graphql/span/dao/SpanLogEventResponseConverter.java [hunk 6]
```diff
@@ -54,18 +54,37 @@
...
         .convert(request, logEvent.getAttributesMap())
-        .map(ConvertedLogEvent::new);
+        .map(
+            attributeMap ->
+                new SpanLogEventPair(
+                    logEvent.getAttributesMap().get(foreignIdAttribute).getString(),
+                    new ConvertedLogEvent(attributeMap)));
+  }
+
+  static class SpanLogEventPair {
```
Comment (skjindal93) [inline hypertrace-core-graphql-span-schema/src/main/java/org/hypertrace/core/graphql/span/dao/SpanLogEventResponseConverter.java:80]: Can we have it as a private static class,  with lombok annotations, so that you don't have to specify a constructor?

File: hypertrace-core-graphql-span-schema/src/main/java/org/hypertrace/core/graphql/span/dao/SpanLogEventResponseConverter.java [hunk 7]
```diff
@@ -54,18 +54,37 @@
...
     return Observable.fromIterable(logEventsResponse.getLogEventsList())
         .concatMapSingle(
-            logEventsResponseVar -> this.convert(attributeRequests, logEventsResponseVar))
-        .collect(Collectors.groupingBy(logEvent -> (String) logEvent.attribute(foreignIdAttribute)))
+            logEventsResponseVar ->
+                this.convert(foreignIdAttribute, attributeRequests, logEventsResponseVar))
+        .collect(
+            Collectors.groupingBy(
+                spanLogEventPair -> spanLogEventPair.spanId,
+                Collectors.mapping(v -> v.logEvent, Collectors.toList())))
```
Comment (skjindal93) [inline hypertrace-core-graphql-span-schema/src/main/java/org/hypertrace/core/graphql/span/dao/SpanLogEventResponseConverter.java:62]: A more descriptive name for `v` or even just have `SpanEventLogPair::logEvent`

File: hypertrace-core-graphql-span-schema/src/test/java/org/hypertrace/core/graphql/span/dao/SpanLogEventRequestBuilderTest.java [hunk 8]
```diff
@@ -54,6 +57,8 @@
 
   @Mock private FilterRequestBuilder filterRequestBuilder;
 
+  @Mock AttributeStore attributeStore;
```
Comment (skjindal93) [inline hypertrace-core-graphql-span-schema/src/test/java/org/hypertrace/core/graphql/span/dao/SpanLogEventRequestBuilderTest.java:60]:
```suggestion
  @Mock private AttributeStore attributeStore;
```

File: hypertrace-core-graphql-span-schema/src/test/java/org/hypertrace/core/graphql/span/dao/SpanLogEventRequestBuilderTest.java [hunk 9]
```diff
@@ -152,4 +162,55 @@ void testBuildRequest() {
...
+            .getValue()
+            .getStringArrayList()
+            .stream()
+            .collect(Collectors.toList()));
+    assertEquals(startTime, logEventsRequest.getStartTimeMillis());
+    assertEquals(endTime, logEventsRequest.getEndTimeMillis());
+    assertEquals(
+        Set.of("traceId", "spanId"),
+        logEventsRequest.getSelectionList().stream()
+            .map(v -> v.getColumnIdentifier().getColumnName())
```
Comment (skjindal93) [inline hypertrace-core-graphql-span-schema/src/test/java/org/hypertrace/core/graphql/span/dao/SpanLogEventRequestBuilderTest.java:213]: A more descriptive name for `v`

File: hypertrace-core-graphql-span-schema/src/main/java/org/hypertrace/core/graphql/span/dao/SpanLogEventRequestBuilder.java (hunk 3 above)
Comment (rish691) [reply hypertrace-core-graphql-span-schema/src/main/java/org/hypertrace/core/graphql/span/dao/SpanLogEventRequestBuilder.java]: Renamed `v` at other places as well, will be careful next time on!

File: hypertrace-core-graphql-span-schema/src/main/java/org/hypertrace/core/graphql/span/dao/SpanLogEventRequestBuilder.java (hunk 4 above)
Comment (rish691) [reply hypertrace-core-graphql-span-schema/src/main/java/org/hypertrace/core/graphql/span/dao/SpanLogEventRequestBuilder.java]: This method is removed now (as per Aaron's suggestion).

File: hypertrace-core-graphql-span-schema/src/main/java/org/hypertrace/core/graphql/span/dao/SpanLogEventResponseConverter.java (hunk 2 above)
Comment (rish691) [reply hypertrace-core-graphql-span-schema/src/main/java/org/hypertrace/core/graphql/span/dao/SpanLogEventResponseConverter.java:57]: yeah right spanId filtering would be automatically done, the problem was building the mapping of `spanId -> List<LogEvent>`. Currently the code was first converting LogEvent (gateway) -> LogEvent (gql) and then mapping spanId to logEvents, but now it is possible that spanId attribute would be lost after conversion.

Comment (aaron-steinfeld) [reply hypertrace-core-graphql-span-schema/src/main/java/org/hypertrace/core/graphql/span/dao/SpanLogEventResponseConverter.java:57]: got it - because the attribute map conversion is based on the requested rather than returned data, and we don't save the extra attribute into the request. makes sense.

**Review by aaron-steinfeld — APPROVED** [review]
